mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,180 @@
+
+import numpy as np
+
+from scipy.linalg import inv
+
+from mne import pick_types, pick_info
+from mne.utils import logger, _validate_type
+from mne.parallel import check_n_jobs
+from mne.io import BaseRaw
+from mne.epochs import BaseEpochs
+from mne.evoked import Evoked
+from mne.parallel import parallel_func
+from mne.channels.interpolation import _calc_g, _calc_h
+
+
+def _prepare_G(G, lambda2):
+    G.flat[::len(G) + 1] += lambda2
+    # compute the CSD
+    Gi = inv(G)
+
+    TC = Gi.sum(0)
+    sgi = np.sum(TC)  # compute sum total
+
+    return Gi, TC, sgi
+
+
+def _compute_csd(data, G_precomputed, H, sphere):
+    """compute the CSD"""
+    n_channels, n_times = data.shape
+    x, y, z, r = sphere
+    mu = data.mean(0)[None]
+    Z = data - mu
+    X = np.zeros_like(data)
+    r **= 2
+
+    Gi, TC, sgi = G_precomputed
+
+    Cp2 = np.dot(Gi, Z)
+    c02 = np.sum(Cp2, axis=0) / sgi
+    C2 = Cp2 - np.dot(TC[:, None], c02[None, :])
+    X = np.dot(C2.T, H).T / r
+    return X
+
+
+def compute_current_source_density(inst, picks=None, g_matrix=None,
+                                   h_matrix=None, lambda2=1e-5,
+                                   stiffness=4, n_legendre_terms=50,
+                                   sphere=(0., 0., 0., 0.095), n_jobs=1,
+                                   copy=True):
+    """Current Source Density (CSD) transformation
+
+    Transformation based on spherical spline surface Laplacian.
+
+    .. note:: This function applies an average reference to the data.
+              Do not transform CSD data to source space.
+
+    Parameters
+    ----------
+    inst : instance of Raw, Epochs or Evoked
+        The data to be transformed.
+    picks : np.ndarray | list, shape (n_channels,) | None
+        The picks to be used. Defaults to None.
+    lambda2 : float
+        Regularization parameter, produces smoothnes. Defaults to 1e-5.
+    stiffnes : float
+        stiffnes of the spline. Also referred to as `m`
+        (if g_matrix or h_matrix is provided this will be ignored)
+    n_legendre_terms : int
+        number of Legendre terms to evaluate (if g_matrix or h_matrix
+        is provided this will be ignored)
+    sphere : tuple, (float, float, float, float)
+        The sphere, head-model of the form (x, y, z, r) where x, y, z
+        is the center of the sphere and r is the radius
+    n_jobs : int
+        The number of processes to run in parallel. Note. Only used for
+        Epochs input. Defaults to 1.
+    copy : bool
+        Whether to overwrite instance data or create a copy.
+
+    Returns
+    -------
+    inst_csd : instance of Epochs or Evoked
+        The transformed data. Output type will match input type.
+    """
+
+    _validate_type(inst, (BaseEpochs, BaseRaw, Evoked), 'inst')
+
+    if 'csd' in inst.info['comps']:
+        raise ValueError('CSD already applied, should not be reapplied')
+
+    if copy is True:
+        out = inst.copy()
+    else:
+        out = inst
```

Comment:
```suggestion
        inst = inst.copy() if copy else inst
```
Then only ever operate on `inst` not `out`